### ISSUE TITLE
`Layout/ClassStructure` bad detection of dynamic constants

### ISSUE THREAD

**Issue by marcandre** [rubocop/rubocop#9739]
#7808 attempted to avoid the cop to move dynamic constants that could depend on other code.

It fails to detect that properly:

```ruby
local = 42
A = [1, 2, 3].freeze
B = "answer: #{local}"
C = [local]
D = local > 40 ? 1 : 2
E = A
```

It believes that `A` is dynamic (when it can be moved freely) and that `B`, `C`, `D`, `E` are ok to move, when they are not.

A better criteria would be: should not move if and only if expression refers to a local variable, a method call with no receiver (or `self`), or a constant without scope (or `self::`)

cc/ @tejasbubane 

I'll handle it in my refactor

**Comment by bbatsov** [comment]
Closing in favor of #9738 which tracks related `Layout/ClassStructure` autocorrect issues. Please follow up there.
